### PR TITLE
Add async functionality to docs

### PR DESCRIPTION
**Pull request by euoia**
This pull request adds details about the new handling of `async` middleware to the documentation.

Additionally, I have:
* Updated v5 examples to use arrow functions.
* Updates some v5 examples to demonstrate use of `async` request handlers.
* Expanded the documentation around how route callbacks are used.

One thing I wasn't sure about - I have added the `async` keyword to some handler callbacks that don't actually `await` on anything as a way of demonstrating this new functionality. I wonder whether it would be better to only use `async` callbacks in code that demonstrates use of `await`? In practise, however, I (and many others, likely) will simply mark all their request handlers as async (there's no downside, is there?).

This addresses https://github.com/expressjs/expressjs.com/issues/1313.

### CODE REVIEW

File: _includes/api/en/5x/app-all.md [hunk 1]
```diff
@@ -11,7 +11,7 @@ The following callback is executed for requests to `/secret` whether using
 GET, POST, PUT, DELETE, or any other HTTP request method:
 
 ```js
-app.all('/secret', function (req, res, next) {
+app.all('/secret', async (req, res, next) => {
```
Comment (nfantone) [inline _includes/api/en/5x/app-all.md]: Are we sure we  need to update **all** handlers in the docs to be `async` now? Good ol' sync handlers are still completely fine and supported. IMHO, we should change them only where it makes sense.

File: _includes/api/en/5x/app-use.md [hunk 2]
```diff
@@ -97,31 +97,31 @@ app.use('/abcd', function (req, res, next) {
 This will match paths starting with `/abcd` and `/abd`:
 
 ```js
-app.use('/abc?d', function (req, res, next) {
+app.use('/abc?d', (req, res, next) => {
```
Comment (nfantone) [inline _includes/api/en/5x/app-use.md]: This and all the others paths featuring modifiers no longer work with the latest `router`. We should either remove them from the docs or re-write them to use matching groups and/or params.

```suggestion
app.use('/ab(c?)d', (req, res, next) => {
```

File: _includes/api/en/5x/app-use.md [hunk 3]
```diff
@@ -97,31 +97,31 @@ app.use('/abcd', function (req, res, next) {
...
+app.use('/ab+cd', (req, res, next) => {
   next()
 })
 ```
 
 This will match paths starting with `/abcd`, `/abxcd`, `/abFOOcd`, `/abbArcd`, and so on:
 
 ```js
-app.use('/ab*cd', function (req, res, next) {
+.use('/ab*cd', (req, res, next) => {
```
Comment (nfantone) [inline _includes/api/en/5x/app-use.md]: Accidental `app` removed?

File: _includes/api/en/5x/app-use.md [hunk 4]
```diff
@@ -97,31 +97,31 @@ app.use('/abcd', function (req, res, next) {
...
   next()
 })
 ```
 
 This will match paths starting with `/ad` and `/abcd`:
 
 ```js
-app.use('/a(bc)?d', function (req, res, next) {
+app.use('/a(bc)?d', (req, res, next) => {
```
Comment (nfantone) [inline _includes/api/en/5x/app-use.md]: Side comment (no action needed): somewhat confusingly, this is still very valid and `/a(bc)?d` works as expected with the new `router`.

File: _includes/api/en/5x/router-param.md [hunk 5]
```diff
@@ -80,7 +77,7 @@ const router = express.Router()
 
 // customizing the behavior of router.param()
 router.param(function (param, option) {
```
Comment (nfantone) [inline _includes/api/en/5x/router-param.md]: Did you not turn this particular `function` into a fat arrow for a reason? Or just an omission?

File: _includes/api/en/5x/routing-args.html [hunk 6]
```diff
@@ -26,18 +26,19 @@ <h4> Arguments</h4>
...
 <li>An array of middleware functions.</li>
-<li>A combination of all of the above.</li>
+<li>Any combination of the above.</li>
 </ul>
 <p>
-You can provide multiple callback functions that behave just like middleware, except
-that these callbacks can invoke <code>next('route')</code> to bypass
-the remaining route callback(s). You can use this mechanism to impose pre-conditions
-on a route, then pass control to subsequent routes if there is no reason to proceed with the current route.
+The callbacks are invoked in the order in which they are provided and behave in the same way as middleware. In order for the next callback to be invoked, the current one must call <code>next()</code>. If a callback doesn't call <code>next()</code> and doesn't respond to the client, the request will hang.
```
Comment (nfantone) [inline _includes/api/en/5x/routing-args.html]:
```suggestion
The callbacks are invoked in the order in which they are provided and behave in the same way as middleware. In order for the next callback to be invoked, the current one must call <code>next()</code>. If a callback doesn't call <code>next()</code> or doesn't respond to the client, the request will hang.
```

File: en/guide/migrating-5.md [hunk 7]
```diff
@@ -126,6 +126,20 @@ In Express 4.7 and Express 5 onwards, the query parser option can accept `false`
 
 <h3>Improvements</h3>
 
+<h4 id="async.request.handlers">async request handlers</h4>
+
+Request handlers and middleware can now be be `async` functions, and they will behave in the same way as non-async handlers when an error occurs.
+
+In Express 4, it was necessary to explicitly handle rejected promises. Without handling the rejection, Node.js would exit and the client would receive an empty response. In Express 5, when middleware returns a rejected promise (for example, by using `await` inside an `async` request handler), Express behaves the same way as when an uncaught error is thrown: That is:
```
Comment (nfantone) [inline en/guide/migrating-5.md]: > Without handling the rejection, Node.js would exit

Probably being too specific here, but Node.js didn't traditionally terminate the process on unhandled promises. Instead, it would print out an `UnhandledPromiseRejectionWarning`

File: en/guide/migrating-5.md [hunk 8]
```diff
@@ -126,6 +126,20 @@ In Express 4.7 and Express 5 onwards, the query parser option can accept `false`
...
+
+1. The client receives an HTTP 500 error,
+
+2. The client is shown provided a stack trace (except when the `NODE_ENV` environment variable is `production`),
```
Comment (nfantone) [inline en/guide/migrating-5.md]:
```suggestion
2. The client is shown the error stack trace (except when the `NODE_ENV` environment variable is `production`),
```

**Review by nfantone — COMMENTED** [review]
Good stuff. Left some comments that I feel we should address before we hear from @dougwilson et al.

File: _includes/api/en/5x/app-all.md (hunk 1 above)
Comment (euoia) [reply _includes/api/en/5x/app-all.md]: Yes, agreed, I will revert changes to `async` where there's no `await` in the handler body.

File: _includes/api/en/5x/router-param.md (hunk 5 above)
Comment (euoia) [reply _includes/api/en/5x/router-param.md]: No reason. Fixed with 81460a22.

File: _includes/api/en/5x/routing-args.html (hunk 6 above)
Comment (euoia) [reply _includes/api/en/5x/routing-args.html]: Hmm, I guess given that it's hard to understand it might be worth rewording to make things even more explicit, for example:

Every callback must either send a response, or call `next()`, allowing one of the subsequent callbacks to send a response. If execution of the series of callbacks ends without sending a response to the client (or erroring), the request will hang.

What do you think?

File: en/guide/migrating-5.md (hunk 7 above)
Comment (euoia) [reply en/guide/migrating-5.md]: You're right.

I guess I can reword to say:

If the rejection was not handled, the client would receive an empty response (as opposed to an error), and as of Node 15 the `UnhandledPromiseRejectionWarning` would throw an error.

What do you think? Still too specific?

File: _includes/api/en/5x/routing-args.html (hunk 6 above)
Comment (nfantone) [reply _includes/api/en/5x/routing-args.html]: Perfect 👍🏼.

File: _includes/api/en/5x/app-all.md (hunk 1 above)
Comment (nfantone) [reply _includes/api/en/5x/app-all.md]: ..._aand_ just noted that you actually addressed this in a comment above. Apologies for missing that 🙏🏼.

File: _includes/api/en/5x/routing-args.html (hunk 6 above)
Comment (euoia) [reply _includes/api/en/5x/routing-args.html]: Fixed with ba717a8d.

File: _includes/api/en/5x/routing-args.html [hunk 9]
```diff
@@ -26,18 +26,19 @@ <h4> Arguments</h4>
...
 </ul>
 <p>
-You can provide multiple callback functions that behave just like middleware, except
-that these callbacks can invoke <code>next('route')</code> to bypass
-the remaining route callback(s). You can use this mechanism to impose pre-conditions
-on a route, then pass control to subsequent routes if there is no reason to proceed with the current route.
+The callbacks are invoked in the order in which they are provided and behave in the same way as middleware. In order for the next callback to be invoked, the current one must call <code>next()</code>. Every callback must either send a response, or call next(), allowing one of the subsequent callbacks to send a response. If execution of the series of callbacks ends without sending a response to the client (or erroring), the request will hang.
+</p>
+<p>
+Throwing an error or returning a rejected promise in a middleware function will cause all further route callbacks to be skipped. This allows a middleware function to be defined using the <code>async</code> keyword and to <code>await</code> on promises.</p>
```
Comment (xjamundx) [inline _includes/api/en/5x/routing-args.html]: I think I'd prefer to call these "async functions" instead of "defined using the async keyword" so like:

> This allows middleware to be defined using async functions

 And then maybe include as part of it some kind of explanation like this: 

> Whenever `await` is called on a promise that is later rejected, the middleware chain will be broken and the default error handler will be called with the rejected value, similar to calling `return next(err)` from inside a middleware function. 

I don't know if it needs to be called out explicitly but because of this architecture it's not necessary to wrap every express middleware or route `await` in `try/catch` which is common in most examples.

File: en/guide/migrating-5.md [hunk 10]
```diff
@@ -126,6 +126,20 @@ In Express 4.7 and Express 5 onwards, the query parser option can accept `false`
 
 <h3>Improvements</h3>
 
+<h4 id="async.request.handlers">async request handlers</h4>
+
+Request handlers and middleware can now be be `async` functions, and they will behave in the same way as non-async handlers when an error occurs.
```
Comment (xjamundx) [inline en/guide/migrating-5.md]: Do you mean "when an error is thrown"?

File: _includes/api/en/5x/routing-args.html (hunk 9 above)
Comment (xjamundx) [reply _includes/api/en/5x/routing-args.html]: Okay that last part is covered pretty well below, sorry!

File: en/guide/migrating-5.md (hunk 7 above)
Comment (xjamundx) [reply en/guide/migrating-5.md]: Is there enough documentation around the default error handler that you can simply reference that it's the same behavior as if you throw an error or call `next(err)` inside of a route or middleware function? Instead of getting into all of these details, not the the details are bad!

Comment (nfantone) [reply en/guide/migrating-5.md]: That's not a bad idea. Unsure about the details on the error handling docs, though.

File: en/guide/migrating-5.md (hunk 10 above)
Comment (nfantone) [reply en/guide/migrating-5.md]: FWIW, I personally think they both read pretty similar.

File: _includes/api/en/5x/app-all.md (hunk 1 above)
Comment (euoia) [reply _includes/api/en/5x/app-all.md]: Should be fixed with 356be22.

File: en/guide/migrating-5.md (hunk 7 above)
Comment (euoia) [reply en/guide/migrating-5.md]: There is documentation around error handling here: http://expressjs.com/en/guide/error-handling.html

It covers the Express 5 behaviour, so I think I can link to it. I'll try to reword to make it a bit less verbose.

File: en/guide/migrating-5.md (hunk 10 above)
Comment (euoia) [reply en/guide/migrating-5.md]: @xjamundx are you happy to resolve this without a change? My intention behind using `when an error occurs` was just to try to keep things simple. I didn't want to confuse the reader by implying they might need to throw the error from their code (as opposed to from some code they're calling). Errors may also come from Node.js (not sure if they would be considered `thrown`).

File: _includes/api/en/5x/routing-args.html (hunk 9 above)
Comment (euoia) [reply _includes/api/en/5x/routing-args.html]: @xjamundx thanks for reviewing this. I've pushed an update with b369599 that attempts to simplify the wording, please let me know what you think.

File: en/guide/migrating-5.md (hunk 7 above)
Comment (euoia) [reply en/guide/migrating-5.md]: I've simplified the wording of the whole block given that error handling is well covered in the error handling guide with 5f3ea21. Please let me know if what you think.

Comment (xjamundx) [reply en/guide/migrating-5.md]: Yeah I think the simplified version works well

File: _includes/api/en/5x/routing-args.html (hunk 9 above)
Comment (xjamundx) [reply _includes/api/en/5x/routing-args.html]: Yeah that's good. I think there's a meta discussion about what's a "route" vs a "route callback" vs a "middleware function", but I don't think we need to figure that out right now. This is good

File: en/guide/migrating-5.md (hunk 10 above)
Comment (xjamundx) [reply en/guide/migrating-5.md]: Here's the case I was concerned wasn't explained properly:

```js
app.get('/data', (req, res) => {
    DB.getData().then(data => { 
        res.send(data.property.that.does.not.exist)); // uncaught exception probably
    });
});
```
vs

```js
app.get('/data', async (req, res) => {
    const data = await DB.getData();
    res.send(data.property.that.does.not.exist); // caught in the normal way
});
```

> Request handlers and middleware can now be be `async` functions, and they will behave in the same way as non-async handlers when an error occurs.

Do you see how it's not really handled the same way. The error before isn't really caught (without domains or async local storage or whatever the way is).

Would any of this additional detail be helpful?

> Request handlers and middleware can now be be `async` functions. When an error is thrown or a rejected promise is `await`ed  inside of an async request handler, those errors will be caught and passed to the default error handler as if calling `next(err)`.

File: en/guide/migrating-5.md (hunk 7 above)
Comment (euoia) [reply en/guide/migrating-5.md]: Will resolve. If you have any more comments on this section let me know.

File: _includes/api/en/5x/routing-args.html (hunk 9 above)
Comment (euoia) [reply _includes/api/en/5x/routing-args.html]: I completely agree. It would be nice if somewhere (maybe the glossary) explains the differences between route callback and middleware - it would also be great if we used consistent vocabulary throughout the documentation. Do you know what the differences are? The only documented difference I'm aware of is that a route callback can call `next('route')` to skip the remaining route callbacks.

File: en/guide/migrating-5.md (hunk 10 above)
Comment (euoia) [reply en/guide/migrating-5.md]: I think that's a good improvement. I've used your text as provided in 500e0408. If you're happy, thumbs up and I'll resolve.

File: _includes/api/en/5x/routing-args.html (hunk 9 above)
Comment (xjamundx) [reply _includes/api/en/5x/routing-args.html]: I think it's simply detected in the code base based on the number of arguments your handler includes. If it's 3 it's a middleware, 4 it's an error handler. Definitely something to call out clearly somewhere.

**Comment by xjamundx** [comment]
What else is needed on this. Everything I looked at now looks great 👍🏻

**Review by cityturbo — APPROVED** [review]

**Comment by xjamundx** [comment]
I was shocked to se these docs missing from the express 5 beta launch. Hope they manage to land soon. As far as I'm concerned this is the main reason for people to upgrade :D

**Comment by dougwilson** [comment]
Yep. Only got so much time, haha. I got them rebased and stuff, and will be landing tonight when off work.

**Comment by dougwilson** [comment]
Well, dang it, it looks like `app.param` doesn't support `async` in beta.1. I'll look into that. @euoia would you be ok with me removing those changes to land the rest, or would you want to split it up? Looks like the migration section should probably have a link added to the TOC on that page as well.

**Comment by euoia** [comment]
Hey @dougwilson, thanks for looking at this and congrats on getting beta.1 out!

Looking at it again, it was a very presumptuous to change app.param to use an `async` function without actually checking whether it works. My understanding is that by using an `async` callback, the callback itself will return a promise (instead of undefined). I am curious about why app.param cares about the return type from its callback? I don't see anything in the [docs](http://expressjs.com/en/5x/api.html#app.param) about the return type of `app.param` being significant.

In any case, by all means, modify the PR as required to press on! 🎉

**Review by dougwilson — APPROVED** [review]
i distilled it down to just the beta.1 change. we'll get the rest landed as it arrives
